src/v3/src/util/webauthnUtils.ts

```diff
@@ -96,3 +121,40 @@ export const webAuthNAuthenticationHandler: WebAuthNAuthenticationHandler = asyn
   // for some reason generic remediator does not allow/expect id field and fails when passed
   return { credentials: omit(OktaAuth.webauthn.getAssertion(credentials), ['id']) };
 };
+
+const challengeDataToCredentialRequestOptions = (challengeData: WebAuthNChallengeDataWithUserVerification): PublicKeyCredentialRequestOptions => ({
+  ...challengeData,
+  challenge: strToBin(challengeData.challenge),
+});
+
+function isAuthenticatorAssertionResponse(response: AuthenticatorAssertionResponse | AuthenticatorResponse): response is AuthenticatorAssertionResponse {
+  return (response as AuthenticatorAssertionResponse).userHandle instanceof ArrayBuffer;
+}
+
+export const webAuthNAutofillActionHandler = async (challengeData: WebAuthNChallengeDataWithUserVerification, abortController: AbortController): Promise<WebAuthNAutofillUICredentials | undefined> => {
+  // if the browser doesn't support Passkey autofill and AbortController, no action needs to be taken
+  // as there are other steps the user can take to proceed
+  const supportsPasskeyAutofill = await isPasskeyAutofillAvailable();
+  const supportsAbortController = typeof AbortController !== 'undefined';
+  if (supportsPasskeyAutofill && supportsAbortController) {
+    try {
+      const credential = await navigator.credentials.get({
+        mediation: 'conditional',
+        publicKey: challengeDataToCredentialRequestOptions(challengeData),
+        signal: abortController.signal,
+      }) as PublicKeyCredential;
+
+      if (isAuthenticatorAssertionResponse(credential.response)) {
+        const credentials: WebAuthNAutofillUICredentials = {
+          clientData: binToStr(credential.response.clientDataJSON),
+          authenticatorData: binToStr(credential.response.authenticatorData),
+          signatureData: binToStr(credential.response.signature),
+          userHandle: binToStr(credential.response.userHandle as ArrayBuffer),
+        };
+        return credentials;
+      }
+    } catch {
+      // TODO: OKTA-814898 - add appropriate error handling
```

Comment:
Ok thanks to @glenfannin-okta for pointing out that we can let errors propagate out of this function (or catch and rethrow as needed). Then we should do a `catch` _in the component_ when `getCredentials` is called. At that point we can then set the widget messages using the setter available on the widget `setMessage`.

Reply:
this makes sense, thank you